# Partial Derivatives of a Cylinder's Volume


The volume of a cylinder is described by $v = \pi r^2 l$ where $r$ is the radius and $l$ the length.

In [1]:
import torch
import math # for pi

In [2]:
def cylinder_volume(my_r, my_l):
  return math.pi * my_r**2 * my_l

Assume $r=3$ and $l=5$.

In [3]:
r = torch.tensor(3.).requires_grad_()
r

tensor(3., requires_grad=True)

In [4]:
l = torch.tensor(5.).requires_grad_()
l

tensor(5., requires_grad=True)

In [5]:
v = cylinder_volume(r, l) # forward pass
v

tensor(141.3717, grad_fn=<MulBackward0>)

In [6]:
v.backward() # autodiff

$\frac{\partial v}{\partial l} = \pi r^2$ at $l=5$

In [7]:
l.grad

tensor(28.2743)

In [8]:
math.pi * 3**2 # prove the result above

28.274333882308138

This means that at $r = 3$, a change in $l$ by one unit corresponds to a change in $v$ of 28.27$\text{m}^3$.

In [9]:
cylinder_volume(3,6)

169.64600329384882

In [10]:
cylinder_volume(3,6) - cylinder_volume(3,5) # divide by l=1 to restore scale & diff units

28.274333882308127

In [11]:
cylinder_volume(3,7) - cylinder_volume(3,6) # divide by l=1 to restore scale & diff units

28.274333882308156

$\frac{\partial v}{\partial r} = 2 \pi r l$ at $r=3$

In [12]:
r.grad # autodiff

tensor(94.2478)

In [13]:
2 * math.pi * 3 * 5

94.24777960769379

$r$ is included in the partial derivative so adjusting it affects the scale of its impact on $v$. Although it's our first example in this notebook, it is typical in calculus for the derivative only to apply at an infinitesimally small $\Delta r$. The smaller the $\Delta r$, the closer to the true $\frac{\partial v}{\partial r}$. E.g., at $\Delta r = 1 \times 10^{-6}$

In [14]:
delta = 1e-6

In [15]:
(cylinder_volume(3+delta, 5) - cylinder_volume(3, 5)) / delta # divide by delta to restore scale & diff units

94.24779531741478